### DISCUSSION TITLE
Serializing filter and filter options

### DISCUSSION THREAD

**Discussion by nahasco** [encode/django-rest-framework#9004] [Question & Answer]
What is the way to serialize filter options and return it with the response of the generic view for example?

I know how to apply filters and send requests with filtering but I cant seem to find any way of serializing the filter options that can be selected in the frontend to make the filtering requests.

**Comment by auvipy** [comment]
care to share code example and API request / response you are willing to achieve?

**Reply by nahasco** [reply]
I did not share my code because I am looking for a global solution. Here is my code nevertheless:

models.py
```
class Halaqa(models.Model):
    institution = models.ForeignKey(Institution, on_delete=models.CASCADE, null=True)
    masjid = models.ForeignKey(Masjid, on_delete=models.CASCADE, null=True)
    title = models.CharField(max_length=25)
    teacher = models.ForeignKey(
        Teacher, on_delete=models.PROTECT, related_name="halaqa"
    )
    students = models.ManyToManyField(Student, blank=True)
    edulevel = models.ForeignKey(
        EduLevel, on_delete=models.PROTECT, related_name="halaqa"
    )
    session = models.ForeignKey(
        Session, on_delete=models.PROTECT, related_name="halaqa"
    )
    days = models.ManyToManyField(Day)
    status = models.IntegerField(choices=STATUS_CHOICES)

    def __str__(self):
        return f"{self.id} - {self.title}"
```
        

views.py
```
class HalaqaList(generics.ListCreateAPIView):
    queryset = Halaqa.objects.all()
    serializer_class = HalaqaSerializer
    filterset_fields = ['masjid', 'edulevel']
    filter_backends = [filters.DjangoFilterBackend]


    def get_queryset(self):
        user = self.request.user
        return super().get_queryset().filter(institution=user.institution)
```

Response
```
HTTP 200 OK
Allow: GET, POST, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "links": {
        "next": null,
        "previous": null
    },
    "current_page": 1,
    "count": 2,
    "total_pages": 1,
    "results": [
        {
            "id": 1,
            "title": "الحلقة الاولى",
            "teacher": {
                "first_name": "محمد",
                "last_name": "احمد"
            },
            "masjid": "مسجد السلام",
            "edulevel": "ابتدائي",
            "status": "active",
            "session": "العصر",
            "days": [
                "الاحد",
                "الخميس"
            ],
            "students_count": 0
        },
        {
            "id": 2,
            "title": "الحلقة الثانية",
            "teacher": {
                "first_name": "محمد",
                "last_name": "احمد"
            },
            "masjid": "مسجد السلام",
            "edulevel": "ابتدائي",
            "status": "active",
            "session": "العصر",
            "days": [
                "الاثنين",
                "الثلاثاء"
            ],
            "students_count": 0
        }
    ]
}
```
        
        
What I am trying to achieve is to develop filter controls in the frontend, I am looking for a way to serialize the available filters based on the model fields and send them with the response of the view. I think this functionality may already be implemented in DRF since we see "Filters" button in the browsable API with filter controls, which is very similar to what I am trying to achieve.

Any help or guidance is highly appreciated!

**Reply by auvipy** [reply]
did you check the docs here? https://www.django-rest-framework.org/api-guide/filtering/ ? you can also try [django-rest](https://django-rest-framework-json-api.readthedocs.io/en/stable/usage.html#filter-backends)

**Reply by nahasco** [reply]
Yes, I did, but I could not find an explanation or mention of how to implement this. In case I missed it, I would appreciate it if you guide me.

**Comment by nahasco** [comment]
Anyone got a solution for this?

**Reply by 4bdulll4** [reply]
You could use the **OPTIONS** method to return a list of valid filters that the backend API supports. Form actions has already been implemented using metadata classes. You can achieve this by overriding the metadata class and define the class either in the **ViewSet** class or as global settings.
Please refer to [metadata classes](https://www.django-rest-framework.org/api-guide/metadata/) in the documentation and check if it serves your purpose.

**Reply by nahasco** [reply]
I tried using the `rest_framework.metadata.SimpleMetadata` and it does not include the filterset fields in the OPTIONS. Any idea how I would include the filter fields/controls in the metadata?

**Reply by nahasco** [reply]
I would appreciate if you provide an example
